src/app/SOCestimator/SOCestimator.c

```diff
@@ -0,0 +1,134 @@
+#include <stdbool.h>
+
+#include "common_macros.h"
+#include "BatteryCharacteristics.h"
+#include "SOCestimator.h"
+
+#define CELL_CAPACITY_Ah 6.550
+#define BATTERY_CAPACITY_Ah (NUM_SERIES_CELLS * CELL_CAPACITY_Ah)
+
+int SOC_raw;
+int SOC_limit;
+
+/**
+ * Use the average series cell voltage and ambient temperature to determine a maximum SOC.
+ * Voltage thresholds determined (shittily) by eyeballing voltage/discharge % graph on the cell datasheet.
+ * avg_V [in] - average series cell voltage
+ * ambient_temp_C [in] - ambient temperature
+ */
+static int SOC_limit_from_voltage(float avg_V, float ambient_temp_C)
+{
+    // determine which temperature range the ambient temp falls into
+    // saturate to safe values
+    int temp_index = SAT((ambient_temp_C + .5) / 10, 0, NUM_TEMP_RANGES - 1);
+    
+    if (avg_V < SOC_25_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 25;
+    }
+    else if (avg_V < SOC_50_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 50;
+    }
+    else if (avg_V < SOC_75_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 75;
+    }
+    else
+    {
+        return 100;
+    }
+}
+
+/**
+ * Calculate the change in SOC (delta SOC) resulting from a current held over an elapsed time.
+ * current_A [in] - current held
+ * time_elapsed_ms [in] - time elapsed
+ */
+static inline float calculate_dSOC(float current_A, unsigned int time_elapsed_ms)
+{
+    return (current_A * time_elapsed_ms * 0.001) / BATTERY_CAPACITY_Ah;
+}
+
+// SOC waits for an average voltage reading to determine starting SOC.
+// This flag stays false until this first average voltage reading is recieved.
+// Current readings up to this point are aggregated and applied to the starting SOC model
+// as soon as the first voltage is received.
+bool starting_soc_calculated = false;
+// keep track of current and time expired prior to the initial SOC estimate
+float pre_init_current_A = 0.0;
+
+void SOCestimator_init(void)
+{
+    SOC_raw = 100;
+    SOC_limit = 100;
+    starting_soc_calculated = false;
+}
+
+/**
+ * Use the instantaneous current to update the coulomb counting SOC model.
+ * current [in] - the instantaneous current (negative for discharge, positive for charging)
+ */
+void SOCestimator_coulomb_count_update_1kHz(float current_A)
+{
+    if (starting_soc_calculated)
+    {
+        SOC_raw = SOC_raw + calculate_dSOC(current_A, 1);
```

Comment:
Are we planning on doing any signal processing on this current value before its passed in? I ask because I imagine the analog input will have a lot of noise on it, and that could really mess things up. Especially if you just take a sample and pass it through.

Reply:
Does this calculation rely on the current measurement being negative? Or else SOC wouldn't go down?

Reply:
yes current must be negative. There will be signal processing but that belongs in CurrentSense